core/engine/src/builtins/map/mod.rs

```diff
@@ -549,6 +555,111 @@ impl Map {
+        };
+
+        // 4. For each Record { [[Key]], [[Value]] } p of M.[[MapData]], do
+        if let Some(existing) = map.get(&key) {
+            // a. If p.[[Key]] is not empty and SameValue(p.[[Key]], key) is true, return p.[[Value]].
+            return Ok(existing.clone());
+        }
+
+        // 5. Let p be the Record { [[Key]]: key, [[Value]]: value }.
+        // 6. Append p to M.[[MapData]].
+        map.insert(key, value.clone());
+        // 7. Return value.
+        Ok(value.clone())
+    }
+
+    /// `Map.prototype.getOrInsertComputed(key, callback)`
+    ///
+    /// If the key exists in the Map, returns the existing value.
+    /// Otherwise computes a new value by calling `callback` with the key,
+    /// inserts it into the Map, and returns it.
+    ///
+    /// More information:
+    ///  - [Upsert proposal reference][spec]
+    ///  - [MDN documentation][mdn]
+    ///
+    /// [spec]: https://tc39.es/proposal-upsert/#sec-map.prototype.getOrInsertComputed
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/getOrInsertComputed
+    pub(crate) fn get_or_insert_computed(
+        this: &JsValue,
+        args: &[JsValue],
+        context: &mut Context,
+    ) -> JsResult<JsValue> {
+        let key = args.get_or_undefined(0);
+        let callback = args.get_or_undefined(1);
+
+        // 1. Let M be the this value.
+        let map_obj = this.as_object();
+        // 2. Perform ? RequireInternalSlot(M, [[MapData]]).
+        let mut map = map_obj
+            .as_ref()
+            .and_then(|obj| obj.downcast_mut::<OrderedMap<JsValue>>())
+            .ok_or_else(|| JsNativeError::typ().with_message("`this` is not a Map"))?;
+
+        // 3. If IsCallable(callbackfn) is false, throw a TypeError exception.
+        let callback = callback.as_callable().ok_or_else(|| {
+            JsNativeError::typ().with_message(format!(
+                "Map.prototype.getOrInsertComputed: {} is not a function",
+                callback.display()
+            ))
+        })?;
+
+        // 4. Set key to CanonicalizeKeyedCollectionKey(key).
+        let key = match key.variant() {
+            // If key is -0𝔽, set key to +0𝔽.
+            JsVariant::Float64(r) if r.is_zero() => JsValue::new(0),
+            _ => key.clone(),
+        };
+
+        // 5. For each Record { [[Key]], [[Value]] } p of M.[[MapData]], do
+        if let Some(existing) = map.get(&key) {
+            // a. If p.[[Key]] is not empty and SameValue(p.[[Key]], key) is true, return p.[[Value]].
+            return Ok(existing.clone());
+        }
+
+        // 6. Let value be ? Call(callback, undefined, « key »).
+        // 7. Let p be the Record { [[Key]]: key, [[Value]]: value }.
+        let value = callback.call(&JsValue::undefined(), std::slice::from_ref(&key), context)?;
+        // 8. Append p to M.[[MapData]].
+        map.insert(key.clone(), value.clone());
+        // 9. Return value.
+        Ok(value)
```

Comment:
The comments here seem to diverge from the spec: https://tc39.es/proposal-upsert/#sec-map.prototype.getOrInsertComputed

BTW, everytime you call a JS function, you _have_ to drop the `RefMut` instance since the callback may use it, leading to a panic (since you cannot mutably reference an object twice). There are 2 test262 that are failing with panics:

- `test262/test/built-ins/Map/prototype/getOrInsertComputed/check-state-after-callback-fn-throws.js`
- `test262/test/built-ins/Map/prototype/getOrInsertComputed/overwrites-mutation-from-callbackfn.js`

A good reference on what to do is in `Set::for_each()`: https://github.com/boa-dev/boa/blob/main/core/engine/src/builtins/set/mod.rs#L494-L523